glambie/data/data_catalogue.py

```diff
@@ -152,6 +168,96 @@ def datasets_are_same_unit(self):
+
+        Parameters
+        ----------
+        remove_trend : bool, optional
+            if set to True, the trend over a shared time period is removed, by default True
+        add_trend_after_averaging : bool, optional
+            this will add the average trends from all the input solutions in the catalogue to the averaged solution
+            this flag is only active when remove_trend is set to True.
+            by default False
+        out_data_group : GlambieDataGroup, optional
+            data group to be assigned to combined output Timeseries metadata,
+            by default GLAMBIE_DATA_GROUPS["consensus"]
+
+
+        Returns
+        -------
+        Tuple[Timeseries, DataCatalogue]
+            1. Timeseries object, containing the new combined solution
+            2. DataCatalogue, containing the catalogue the operation was performed on.
+               For remove_trend set to False, this is an exact copy of self.
+               For remove_trend set to True, this contains the altered datasets with the trends removed
+        """
+
+        if not self.datasets_are_same_unit():
+            warnings.warn("Warning: Trends within catalogue are not all the same unit. \
+                This might lead to false results.")
+
+        # merge all dataframes
+        catalogue_dfs = [ds.data.as_dataframe() for ds in self.datasets]
+
+        data_catalogue_out = self.copy()  # DataCatalogue object for returning
+
+        # remove trend / calculate beta instead of B
+        if remove_trend:
+            means_over_period = []  # keep track of the means for adding back in case add_trend_after_averaging=True
+            start_ref_period = np.max([df.start_dates.min() for df in catalogue_dfs])
+            end_ref_period = np.min([df.end_dates.max() for df in catalogue_dfs])
+
+            if not start_ref_period < end_ref_period:
+                warnings.warn("Warning when removing trends. No common period detected.")
+            for idx, df in enumerate(catalogue_dfs):
+                df_sub = df[(df["start_dates"] >= start_ref_period) & (df["end_dates"] <= end_ref_period)]
+                df["changes"] = df["changes"] - df_sub["changes"].mean()
+                data_catalogue_out.datasets[idx].data.changes = np.array(df["changes"])
+                means_over_period.append(df_sub["changes"].mean())
+
+        df = reduce(lambda left, right: left.merge(right, how="outer", on=["start_dates", "end_dates"]), catalogue_dfs)
+        df = df.sort_values(by="start_dates")
+        start_dates, end_dates = np.array(df["start_dates"]), np.array(df["end_dates"])
+        mean_changes = np.array(df[df.columns.intersection(
+            df.filter(regex=("changes*")).columns.to_list())].mean(axis=1))
+
+        # UNCERTAINTIES -- more information is in GlaMBIE Assessment Algorithm document
+        # propagate observational errors
+        sigma_observational_errors = ((df.filter(regex=("errors*"))**2).sum(axis=1))**0.5
```

Comment:
I can see from this math that yes, you are working with standard errors.

Ideally we'd do some renames to ensure that we never say "error" or "uncertainty" in isolation in code, as these are ambiguous terms.

I'm resonably sure that the approach here will result in an over-estimate, though, which is fine. I'd only worry if I saw something that would be likely to lead to an underestimate of the error.

Reply:
Renaming to "uncertainties". In terms of the algorithm, there is no need to check as this is all defined by the GlaMBIE team plus advisory board... Except for a sanity check if the implementation makes sense :)